File: test/FSharp.ActiveLogin.Identity.Swedish.Test/SwedishPersonalIdentityNumber_Create.cs
using System;
using Xunit;

namespace ActiveLogin.Identity.Swedish.Test
{
    /// <remarks>
    /// Tested with offical test Personal Identity Numbers from Skatteverket:
    /// https://skatteverket.entryscape.net/catalog/9/datasets/147
    /// </remarks>
    public class SwedishPersonalIdentityNumber_Create
    {
        [Theory]
        [InlineData(-1, 01, 01, 239, 2)]
        [InlineData(int.MaxValue, 01, 01, 239, 2)]
        public void Throws_When_Invalid_Year(int year, int month, int day, int birthNumber, int checksum)
        {
            var ex = Assert.Throws<ArgumentOutOfRangeException>(() => SwedishPersonalIdentityNumber.Create(year, month, day, birthNumber, checksum));
            Assert.Contains("Invalid year.", ex.Message);
        }

        [Theory]
        [InlineData(2018, 0, 01, 239, 2)]
        [InlineData(2018, 13, 01, 239, 2)]
        public void Throws_When_Invalid_Month(int year, int month, int day, int birthNumber, int checksum)
        {
            var ex = Assert.Throws<ArgumentOutOfRangeException>(() => SwedishPersonalIdentityNumber.Create(year, month, day, birthNumber, checksum));
            Assert.Contains("Invalid month.", ex.Message);
        }

        [Theory]
        [InlineData(2018, 01, 0, 239, 2)]
        [InlineData(2018, 01, 32, 239, 2)]
        [InlineData(2018, 02, 30, 239, 2)]
        public void Throws_When_Invalid_Day(int year, int month, int day, int birthNumber, int checksum)
        {
            var ex = Assert.Throws<ArgumentOutOfRangeException>(() => SwedishPersonalIdentityNumber.Create(year, month, day, birthNumber, checksum));
            Assert.Contains("Invalid day of month.", ex.Message);
        }

        [Theory]
        [InlineData(2018, 01, 61, 239, 2)]
        public void Throws_When_Possible_CoOrdinationNumber(int year, int month, int day, int birthNumber, int checksum)
        {
            var ex = Assert.Throws<ArgumentOutOfRangeException>(() => SwedishPersonalIdentityNumber.Create(year, month, day, birthNumber, checksum));
            Assert.Contains("Invalid day of month.", ex.Message);
        }

        [Theory]
        [InlineData(2018, 01, 01, 0, 2)]
        [InlineData(2018, 01, 01, 1000, 2)]
        public void Throws_When_Invalid_BirthNumber(int year, int month, int day, int birthNumber, int checksum)
        {
            var ex = Assert.Throws<ArgumentOutOfRangeException>(() => SwedishPersonalIdentityNumber.Create(year, month, day, birthNumber, checksum));
            Assert.Contains("Invalid birth number.", ex.Message);
        }

        [Theory]
        [InlineData(2018, 01, 01, 239, 3)]
        [InlineData(2018, 01, 01, 239, 4)]
        public void Throws_When_Invalid_Checksum(int year, int month, int day, int birthNumber, int checksum)
        {
            var ex = Assert.Throws<ArgumentException>(() => SwedishPersonalIdentityNumber.Create(year, month, day, birthNumber, checksum));
            Assert.Contains("Invalid checksum.", ex.Message);
        }

        [Theory]
        [InlineData(1899, 09, 13, 980, 1)]
        [InlineData(1999, 08, 07, 239, 1)]
        [InlineData(2000, 01, 02, 239, 1)]
        [InlineData(2018, 01, 01, 239, 2)]
        public void Accepts_Valid_Personal_Identity_Number(int year, int month, int day, int birthNumber, int checksum)
        {
            var personalIdentityNumber = SwedishPersonalIdentityNumber.Create(year, month, day, birthNumber, checksum);
            Assert.Equal(year, personalIdentityNumber.Year);
            Assert.Equal(month, personalIdentityNumber.Month);
            Assert.Equal(day, personalIdentityNumber.Day);
            Assert.Equal(birthNumber, personalIdentityNumber.BirthNumber);
            Assert.Equal(checksum, personalIdentityNumber.Checksum);
        }
    }
}

File: test/ActiveLogin.Identity.Swedish.FSharp.Test/SwedishPersonalIdentityNumber_create.fs
/// <remarks>
/// Tested with offical test Personal Identity Numbers from Skatteverket:
/// https://skatteverket.entryscape.net/catalog/9/datasets/147
/// </remarks>
module ActiveLogin.Identity.Swedish.FSharp.Test.SwedishPersonalIdentityNumber_create

open Swensen.Unquote
open Expecto
open FsCheck
open ActiveLogin.Identity.Swedish.FSharp
open System.Reflection
open ActiveLogin.Identity.Swedish.FSharp.TestData
open PinTestHelpers


[<Tests>]
let tests =
    testList "SwedishPersonalIdentityNumber.create" [
        testProp "roundtrip 12DigitString -> create -> to12DigitString" <| fun (Gen.Pin.Valid12Digit str) ->
            str
            |> Gen.stringToValues
            |> SwedishPersonalIdentityNumber.create
            |> SwedishPersonalIdentityNumber.to12DigitString =! str

        testPropWithMaxTest 20000 "invalid year returns InvalidYear Error" <|
            fun (Gen.Pin.ValidValues (y, m, d, b, c), Gen.InvalidYear invalidYear) ->
                toAction SwedishPersonalIdentityNumber.create (invalidYear, m, d, b, c)
                |> Expect.throwsWithMessage "Invalid year."

        testPropWithMaxTest 20000 "valid year does not return InvalidYear Error" <|
            fun (Gen.Pin.ValidValues (y, m, d, b, c), Gen.ValidYear validYear) ->
                toAction SwedishPersonalIdentityNumber.create (y, m, d, b, c)
                |> Expect.doesNotThrowWithMessage "year"

        testProp "invalid month returns InvalidMonth Error" <|
            fun (Gen.Pin.ValidValues (y, m, d, b, c), Gen.InvalidMonth invalidMonth) ->
                toAction SwedishPersonalIdentityNumber.create (y, invalidMonth, d, b, c)
                |> Expect.throwsWithMessage "Invalid month."

        testProp "valid month does not return InvalidMonth Error" <|
            fun (Gen.Pin.ValidValues (y, m, d, b, c), Gen.ValidMonth validMonth) ->
                toAction SwedishPersonalIdentityNumber.create (y, m, d, b, c)
                |> Expect.doesNotThrowWithMessage "month"

        testProp "invalid day returns InvalidDay Error" <| fun (Gen.Pin.WithInvalidDay (y, m, d, b, c)) ->
            toAction SwedishPersonalIdentityNumber.create (y, m, d, b, c)
            |> Expect.throwsWithMessage "Invalid day of month."

        testProp "valid day does not return InvalidDay Error" <| fun (Gen.Pin.WithValidDay (y, m, d, b, c)) ->
            toAction SwedishPersonalIdentityNumber.create (y, m, d, b, c)
            |> Expect.doesNotThrowWithMessage "day"

        testProp "invalid birth number returns InvalidBirthNumber Error" <|
            fun (Gen.Pin.ValidValues (y, m, d, b, c), Gen.InvalidBirthNumber invalidBirthNumber) ->
                toAction SwedishPersonalIdentityNumber.create (y, m, d, invalidBirthNumber, c)
                |> Expect.throwsWithMessage "Invalid birth number"

        testPropWithMaxTest 3000 "valid birth number does not return InvalidBirthNumber Error" <|
            fun (Gen.Pin.ValidValues (y, m, d, b, c), Gen.ValidBirthNumber validBirthNumber) ->
                toAction SwedishPersonalIdentityNumber.create (y, m, d, validBirthNumber, c)
                |> Expect.doesNotThrowWithMessage "birth"

        testProp "invalid checksum returns InvalidChecksum Error" <|
            fun (Gen.Pin.ValidValues (y, m, d, b, c)) ->
                let invalidChecksums =
                    [ 0..9 ]
                    |> List.except [ c ]

                let withInvalidChecksums =
                    invalidChecksums
                    |> List.map (fun checksum -> (y, m, d, b, checksum))

                withInvalidChecksums
                |> List.iter (fun (y, m, d, b, cs) ->
                    toAction SwedishPersonalIdentityNumber.create (y, m, d, b, cs)
                    |> Expect.throwsWithMessage "Invalid checksum." )

        testCase "fsharp should have no public constructor" <| fun () ->
            let typ = typeof<SwedishPersonalIdentityNumber>
            let numConstructors = typ.GetConstructors(BindingFlags.Public) |> Array.length
            numConstructors =! 0 ]
